File: problems/coding/011-number-of-islands/solution.js
function numIslands(grid) {
  if (!grid.length) return 0;
  const rows = grid.length, cols = grid[0].length; const seen = new Set(); let count = 0;
  const dfs = (r, c) => {
    const k = `${r},${c}`;
    if (seen.has(k) || r < 0 || c < 0 || r >= rows || c >= cols || grid[r][c] !== "1") return;
    seen.add(k); dfs(r+1, c); dfs(r-1, c); dfs(r, c+1); dfs(r, c-1);
  };
  for (let r = 0; r < rows; r++) for (let c = 0; c < cols; c++)
    if (grid[r][c] === "1" && !seen.has(`${r},${c}`)) { count++; dfs(r, c); }
  return count;
}

File: problems/coding/011-number-of-islands/solution.py
def numIslands(grid):
    if not grid: return 0
    rows, cols = len(grid), len(grid[0]); seen = set(); count = 0
    def dfs(r, c):
        if (r, c) in seen or r < 0 or c < 0 or r >= rows or c >= cols or grid[r][c] != "1": return
        seen.add((r, c))
        dfs(r+1, c); dfs(r-1, c); dfs(r, c+1); dfs(r, c-1)
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == "1" and (r, c) not in seen:
                count += 1; dfs(r, c)
    return count
